### PR TITLE
fix(model/sync): remove duplicated index after change column

### PR DESCRIPTION
**Pull request by lext-7**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [ ] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [ ] Have you added new tests to prevent regressions?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [ ] Did you follow the commit message conventions explained in [CONTRIBUTING.md](../CONTRIBUTING.md)?

### Description of change

Remove duplicated indexes after change column in sync.
https://github.com/sequelize/sequelize/issues/7606#issuecomment-364340666

As for alerting a unique column(`alert xxx unique`) will create a unique index, there will be multiple unique indexes for a column after sync with alert multiple times. So it need to find duplicated indexes after change column and remove them.

### CODE REVIEW

File: lib/model.js [hunk 1]
```diff
@@ -1182,6 +1182,22 @@ class Model {
         // Assign an auto-generated name to indexes which are not named by the user
         this.options.indexes = this.QueryInterface.nameIndexes(this.options.indexes, this.tableName);
 
+        const existedIndexes = {};
+        const duplicatedIndexes = [];
+
+        indexes.forEach(index => {
+          const fieldNameStr = index.fields.map(field => field.attribute).join('_');
+          const key = `${index.type}_${index.primary}_${index.unique}_${fieldNameStr}`;
```
Comment (sushantdhiman) [inline lib/model.js]: `nameIndexes` method also doing something similar, to keep names consistent we should share this code

File: test/integration/sequelize.test.js [hunk 2]
```diff
@@ -1018,6 +1018,27 @@ describe(Support.getTestDialectTeaser('Sequelize'), () => {
         .to.be.rejectedWith('Database "cyber_bird" does not match sync match parameter "/$phoenix/"');
     });
 
+    it('won\'t create duplicated unique indexes', function() {
```
Comment (sushantdhiman) [inline test/integration/sequelize.test.js]: Move this test [here](https://github.com/sequelize/sequelize/blob/83dfc5ae9cf671d448690bc3c8a6d0f5ef45a681/test/integration/model/sync.test.js#L133)

Change title `should not create duplicate unique indexes with unique:true`

**Comment by lext-7** [comment]
Sorry, I finally find this way cannot solve this problem perfectly. 

I think `changeColumnQuery()` should be async and return promise, Because it needs to query indexes and constraints. According to existed indexes and constraints, ignore duplicated indexes or constraint and remove those removed indexes or constraints. 

I think removing constraint or index before changing column or after changing column is a hack way for `sync()`. If you use `changeColumn` function directly, error will be raised still.

But changing `changeColumnQuery()` to be async will take a lot of work and conflict with existing api.
What's your opinion?

**Comment by sushantdhiman** [comment]
I agree `changeColumnQuery` needs to take care of existing indexes, can we query for those indexes before hand and pass to `changeColumnQuery` and take care of them

There is one more issue I discovered when working on https://github.com/sequelize/sequelize/pull/9033

For MySQL if we issue `ALTER TABLE <table> CHANGE <col> UNIQUE` even if nothing is change a new index will be created. This is an inherent bug in MySQL

So whenever we try to `sync({ alter: true })` and column with unique constraint will get a new index. Proper way to handle this situation is

1. Fix that bug with MySQL :D 
2. Scan indexes to see if we are creating any duplicate indexes, Which I think is quite some work in itself

You are suggesting (2), you are right it will take some work but this wont be breaking change `changeColumnQuery` (From Query Generator) is a private API on other hand `changeColumn` (From QueryInterface) is public. So there should be no problem changing that query generator method.

Again this is not a critical issue as Sequelize recommend against using `{ alter: true }` in production

**Comment by sushantdhiman** [comment]
Cleanup, closing this PR. If someone wants to takeover please open a new PR

**Pull request closed without merging**